### PR TITLE
Query: Apply search condition optimizations on join conditions

### LINKED ISSUE #4748: Query :: invalid sql generated for query with joins where key selector is a null comparison
query:

```
from instance in ctx.ExamInstances
join answered in ctx.ExamInstanceQuestions
on true equals answered.AnswerId != null
```

generated query is:

```
SELECT [answered].[ExamInstanceQuestionId], [answered].[AnswerId], [answered].[ExamInstanceId], [answered].[QuestionId]
FROM [ExamInstances] AS [instance]
INNER JOIN [ExamInstanceQuestions] AS [answered] ON 1 = [answered].[AnswerId] IS NOT NULL
```

we should be converting the null check into CASE statement, like we do for projections

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #4748 
Apply search condition optimizations on join conditions

### CODE REVIEW

**Comment by smitpatel** [comment]
@anpete - Need to understand why extra order by comes for left join queries.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 1]
```diff
@@ -274,19 +274,45 @@ public virtual Expression VisitSelect(SelectExpression selectExpression)
...
+            var binaryExpression = newExpression as BinaryExpression;
 
-            newExpression
-                = relationalNullsOptimizedExpandingVisitor.IsOptimalExpansion
-                    ? optimizedExpression
-                    : new RelationalNullsExpandingVisitor().Visit(newExpression);
+            if (joinCondition
+                && binaryExpression != null)
+            {
+                var relationalNullsOptimizedExpandingVisitor = new RelationalNullsOptimizedExpandingVisitor();
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: those two are common for both cases - pull them out of the if block

Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: pull this out of the if block

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 2]
```diff
@@ -274,19 +274,45 @@ public virtual Expression VisitSelect(SelectExpression selectExpression)
...
+                var optimizedLeftExpression = relationalNullsOptimizedExpandingVisitor.Visit(binaryExpression.Left);
+
+                optimizedLeftExpression
+                    = relationalNullsOptimizedExpandingVisitor.IsOptimalExpansion
+                        ? optimizedLeftExpression
+                        : new RelationalNullsExpandingVisitor().Visit(binaryExpression.Left);
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: RelationalNullsExpandingVisitor doesn't have state so you can make one of them and reuse it

**Comment by maumar** [comment]
:shipit:  otherwise

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs (hunk 1 above)
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Does it have state? `IsOptimalExpansion`?

**Pull request merged**